src/s390x/zipl.rs

```diff
@@ -33,6 +34,101 @@ pub fn chreipl<P: AsRef<Path>>(dev: P) -> Result<()> {
     Ok(())
 }
 
+/// Secure boot (Secure IPL) includes support of SCSI and ECKD DASD boot devices
+enum Loaddev {
+    Eckd(String),
+    Scsi(String, String, String),
+}
+
+fn parse_lszdev_eckd(line: &str) -> Result<Loaddev> {
+    // ECKD ID looks like: 0.0.5223, we need only last part of it (5223)
+    lazy_static! {
+        static ref REGEX: Regex = Regex::new(r#"[[:digit:]].[[:digit:]].([[:xdigit:]]+)"#).unwrap();
+    }
+    if let Some(cap) = REGEX.captures_iter(line).next() {
+        return Ok(Loaddev::Eckd(cap[1].to_string()));
+    }
+    bail!("bad ECKD id: {}", line);
+}
+
+fn parse_lszdev_zfcp(line: &str) -> Result<Loaddev> {
+    // SCSI ID looks like: 0.0.8000:0x500507630400d1e3:0x4000401d00000000
+    // So here is regex to parse required ids: 8000,500507630400d1e3,4000401d00000000
+    lazy_static! {
+        static ref REGEX: Regex = Regex::new(
+            r#"[[:digit:]].[[:digit:]].([[:xdigit:]]+):0x([[:xdigit:]]+):0x([[:xdigit:]]+)"#
+        )
+        .unwrap();
+    }
+    if let Some(cap) = REGEX.captures_iter(line).next() {
+        return Ok(Loaddev::Scsi(
+            cap[1].to_string(),
+            cap[2].to_string(),
+            cap[3].to_string(),
+        ));
+    }
+    bail!("bad zFCP id: {}", line);
+}
+
+fn parse_lszdev<P: AsRef<Path>>(dev: P) -> Result<Loaddev> {
+    // We don't want to traverse sysfs and do same stuff lszdev does,
+    // so just call it to get required info. Here is sample output:
+    // $ lszdev -c TYPE,ID -n
+    // dasd-eckd    0.0.0190
+    // zfcp-lun     0.0.8007:0x500507630400d1e3:0x4001404c00000000
+    // qeth         0.0.bdd0:0.0.bdd1:0.0.bdd2
+    // generic-ccw  0.0.000c
+    let output = runcmd_output!(
+        "lszdev",
+        "-n",
+        "--columns",
+        "TYPE,ID",
+        "--by-node",
+        dev.as_ref()
+    )?;
+    let (devtype, id) = output
+        .trim()
+        .split_once(' ')
+        .with_context(|| format!("parsing lszdev {output}"))?;
+    match devtype {
+        "dasd-eckd" => parse_lszdev_eckd(id),
+        "zfcp-lun" => parse_lszdev_zfcp(id),
+        _ => bail!("unsupported device: {} id: {}", devtype, id),
+    }
+}
+
+/// Sets zVM Secure Boot (Secure IPL) boot device to `dev`.
+pub fn set_loaddev<P: AsRef<Path>>(dev: P) -> Result<()> {
+    // check if system is zVM guest
+    if !Path::new("/dev/vmcp").exists() {
+        return Ok(());
+    }
```

Comment:
Should we silently accept this? ISTM like we should fail pretty loudly if secure IPL was requested in an invalid context.